In [9]:
import numpy as np 
from sympy import Matrix, symbols, log, simplify
import matplotlib.pyplot as plt
import pandas as pd

Ejercicio, tarea:  

Obtener las funciones discriminantes para las muestras:
$$
\begin{aligned}
& \omega_1=\{(0.5,10.5),(1,12.5),(3,10.5),(3,12.5),(3,14.5),(3,18),(5,18),(5,16),(5,14.5),(5,13)\} \\
& \omega_2=\{(6,9),(8,10),(9,11),(8.5,12),(7,13.5)(8,16)\}
\end{aligned}
$$

Representarlo gráficamente.
$\qquad$ *

In [5]:
def clasif_e(samples):
    # Vector genérico usando Sympy
    X = Matrix([symbols('x' + str(i+1)) for i in range(samples[0].shape[0])])
    print(' Variables : ', X)
    fds = []
    for s in samples:
        m = Matrix(np.mean(s, axis=1))
        fds.append(simplify(X.T*m - (m.T*m)/2))
    return fds

In [3]:
omega1 = np.array([[0.5, 10.5],[1, 12.5],[3, 10.5],[3, 12.5],[3, 14.5],[3, 18],[5, 18],[5, 16],[5, 14.5],[5, 13]]).T
omega2 = np.array([[6, 9],[8, 10],[9, 11],[8.5, 12],[7, 13.5],[8, 16]]).T
z_1 = np.mean(omega1, axis=1)
z_2 = np.mean(omega2, axis=1)
print(z_1)
print(z_2)

[ 3.35 14.  ]
[ 7.75       11.91666667]


$$
\begin{gathered}
z_1=\left\{\binom{3.35}{14}\right\} \\
z_2=\left\{\binom{7.75}{11.9166}\right\}
\end{gathered}
$$

Las funciones discrimantes son:
$$
\begin{gathered}
f d_K(X)=X^T \cdot Z_K-\frac{1}{2} \cdot Z_K^T \cdot Z_K \quad \forall k \in K \\
f d_1(X)=X^T \cdot\binom{3.35}{14}-\frac{1}{2} \cdot\binom{3.35}{14}^T \cdot\binom{3.35}{14} \\
f d_2(X)=X^T \cdot\binom{7.75}{11.9166}-\frac{1}{2} \cdot\binom{7.75}{11.9166}^T \cdot\binom{7.75}{11.9166}
\end{gathered}
$$

Desarrollando las operaciones:
$$
\begin{gathered}
f d_1(X)=3.35 x_1+14 x_2-103.611 \\
f d_2(X)=7.75 x_1+11.9166 x_2-101.033
\end{gathered}
$$

In [10]:
samples = (omega1, omega2)
fds = clasif_e(samples)
print(fds)

 Variables :  Matrix([[x1], [x2]])
[Matrix([[3.35*x1 + 14.0*x2 - 103.61125]]), Matrix([[7.75*x1 + 11.9166666666667*x2 - 101.034722222222]])]


In [11]:
# fd1
display(fds[0])
# fd2
display(fds[1])

Matrix([[3.35*x1 + 14.0*x2 - 103.61125]])

Matrix([[7.75*x1 + 11.9166666666667*x2 - 101.034722222222]])

Tarea: 

Obtener las funciones discriminantes para las muestras:
$$
\begin{aligned}
& \omega_1=\{(0.5,10.5),(1,12.5),(3,10.5),(3,12.5),(3,14.5),(3,18),(5,18),(5,16),(5,14.5),(5,13)\} \\
& \omega_2=\{(6,9),(8,10),(9,11),(8.5,12),(7,13.5)(8,16)\}
\end{aligned}
$$

Nota: Las matrices de covarianzas no son iguales: se obtendrán funciones discriminantes cónicas.
$\qquad$

In [12]:
m_1 = np.mean(omega1, axis=1)
m_2 = np.mean(omega2, axis=1)
print(m_1)
print(m_2)

[ 3.35 14.  ]
[ 7.75       11.91666667]


$$
\begin{gathered}
\vec{m}_1=\binom{3.35}{14} \\
\vec{m}_2=\binom{7.75}{11.9166}
\end{gathered}
$$

Las matrices de covarianzas:
$$
\begin{gathered}
c 11=\frac{1}{10}\left((0.5-3.35)^2+(1-3.35)^2+(3-3.35)^2+\cdots+(5-3.35)^2\right) \\
c 11=2.5025
\end{gathered}
$$
$$
\begin{gathered}
c 12=c 21=\frac{1}{10}((0.5-3.35)(10.5-14)+(1-3.35)(12.5-14)+(3-3.35)(10.5-14)+\cdots+(5-3.35)(13-14)) \\
c 12=2.275 \\
c 22=\frac{1}{10}\left((10.5-14)^2+(12.5-14)^2+(10.5-14)^2+\cdots+(13-14)^2\right) \\
c 22=6.65 \\
C_1=\left(\begin{array}{cc}
2.5025 & 2.275 \\
2.275 & 6.65
\end{array}\right)
\end{gathered}
$$

In [13]:
cov_1 = np.cov(omega1, bias=True)
cov_1

array([[2.5025, 2.275 ],
       [2.275 , 6.65  ]])